### PR TITLE
MetalLB operator enablement in CI

### PR DESCRIPTION
**Pull request by Shilpa-Gokul**
Create a new playbook for MetalLB operator installation, validation and run e2e tests

### CODE REVIEW

File: playbooks/roles/ocp-metallb-operator/tasks/e2e.yml [hunk 1]
```diff
@@ -0,0 +1,1682 @@
...
+
+- name: Validate the controller and speaker pods can be scheduled on the tainted nodes based on the tolerations in metallb CR
+  block:
+  - name: Check the number of nodes running
+    shell: oc get node | grep "Ready" | wc -l
+    register: metallb_nodes
+
+  - name: Taint master and worker nodes
+    shell: |
+      oc adm taint node master-0 master-group=infra:NoSchedule
```
Comment (pravin-dsilva) [inline playbooks/roles/ocp-metallb-operator/tasks/e2e.yml:14]: Can you ensure that the node names come from the cluster as node names may not always remain the same as master-0,worker-0 etc.

File: playbooks/roles/ocp-metallb-operator/tasks/install.yml [hunk 2]
```diff
@@ -0,0 +1,293 @@
+---
+# tasks file for playbooks/roles/ocp-metallb-operator
+
+- name: Check if cluster operators and nodes are healthy
+  include_role:
+    name: check-cluster-health
+
+- name: Clean up
```
Comment (pravin-dsilva) [inline playbooks/roles/ocp-metallb-operator/tasks/install.yml:8]: Can you have a separate file to perform clean up the operator to improve reusability?

**Review by pravin-dsilva — DISMISSED** [review]

**Comment by pravin-dsilva** [comment]
@Shilpa-Gokul  is there a reason why you need to have e2e tests in this repo? There are several tests present 
 for metallb operator in https://github.com/metallb/metallb-operator/tree/main/test/e2e or https://github.com/openshift/openshift-tests-private/blob/master/test/extended/networking/metallb.go. Can these be re-used?

**Comment by yussufsh** [comment]
Do we need a way to combine this with #49? Seems like duplicated efforts to me.

**Comment by AdityaHonkalas** [comment]
Closed PR [#49 ](https://github.com/ocp-power-automation/ocp4-playbooks-extras/pull/49) due to duplication.

File: .idea/vcs.xml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply .idea/vcs.xml]: Removed it

File: playbooks/roles/ocp-metallb-operator/tasks/install.yml (hunk 2 above)
Comment (Shilpa-Gokul) [reply playbooks/roles/ocp-metallb-operator/tasks/install.yml]: Moved to a separate file

**Comment by Shilpa-Gokul** [comment]
> @Shilpa-Gokul is there a reason why you need to have e2e tests in this repo? There are several tests present for metallb operator in https://github.com/metallb/metallb-operator/tree/main/test/e2e or https://github.com/openshift/openshift-tests-private/blob/master/test/extended/networking/metallb.go. Can these be re-used?

@pravin-dsilva Tests from both repos are re-used and testcases that are mentioned in polarion is also covered in this

File: playbooks/roles/ocp-metallb-operator/tasks/e2e.yml [hunk 4]
```diff
@@ -0,0 +1,1707 @@
+---
```
Comment (yussufsh) [inline playbooks/roles/ocp-metallb-operator/tasks/e2e.yml:1]: @Shilpa-Gokul This file is too big to review :(
Can you create reusable tasks and checks if needed?

File: playbooks/roles/ocp-metallb-operator/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+---
+# tasks file for playbooks/roles/ocp-metallb-operator
+- include_tasks: cleanup.yml
+- include_tasks: install.yml
+- include_tasks: e2e.yml
```
Comment (yussufsh) [inline playbooks/roles/ocp-metallb-operator/tasks/main.yml:5]: Please insert a new line at the EOF everywhere you see this symbol.

File: playbooks/roles/ocp-metallb-operator/vars/main.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+---
+# vars file for playbooks/roles/ocp-metallb-operator
+metallb_e2e_env:
+  PATH: "/usr/local/go/bin:{{ ansible_env.PATH }}"
+  KUBECONFIG: "{{ ansible_env.HOME }}/.kube/config"
+  IS_OPENSHIFT: true
+  NAMESPACE: metallb-system
```
Comment (yussufsh) [inline playbooks/roles/ocp-metallb-operator/vars/main.yml:7]: Why are we not using this env everywhere? I see them harcoded.

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: playbooks/roles/ocp-metallb-operator/vars/main.yml (hunk 6 above)
Comment (Shilpa-Gokul) [reply playbooks/roles/ocp-metallb-operator/vars/main.yml:7]: This env file is passed only to the e2e tests from metallb-operator repo.

File: playbooks/roles/ocp-metallb-operator/tasks/cleanup.yml [hunk 7]
```diff
@@ -0,0 +1,99 @@
+---
+# Clean up the existing MetalLB operator resources if any before proceeding with the new deployment
+- name: Clean up
+  block:
+  - name: Get the name of cluster service version
+    shell: oc get clusterserviceversion -n metallb-system | awk '{ if ($1 ~ /metallb/) print $1 }'
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply playbooks/roles/ocp-metallb-operator/tasks/cleanup.yml:6]: Matching the output with a substring so does removing headers make any difference?

File: examples/metallb_vars.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+qe_catalog_source: "qe-app-registry"
+metallb_upgrade_channel: "stable"
+metallb_directory: "/tmp/metallb"
+metallb_golang_tarball: "https://go.dev/dl/go1.20.6.linux-ppc64le.tar.gz"
+metallb_e2e: true
+metallb_e2e_git_repository: "https://github.com/openshift/metallb-operator.git"
+metallb_git_branch: "main"
+metallb_cleanup: true
+metallb_frr_ver: "frr-8"
+metallb_frr_repo: "https://rpm.frrouting.org/repo/{{metallb_frr_ver}}-repo-1-0.el9.noarch.rpm"
```
[the remark this answers is not in the document]
Comment (Shilpa-Gokul) [reply examples/metallb_vars.yml:20]: This is what is being used in the polarion testcases documentation, so I used the same. Should this be changed?

[the remark this answers is not in the document]
Comment (yussufsh) [reply examples/metallb_vars.yml:20]: Just change the name.
```suggestion
metallb_frr_rpm : "https://rpm.frrouting.org/repo/{{metallb_frr_ver}}-repo-1-0.el9.noarch.rpm"
```

File: playbooks/roles/ocp-metallb-operator/tasks/e2e.yml (hunk 4 above)
Comment (Shilpa-Gokul) [reply playbooks/roles/ocp-metallb-operator/tasks/e2e.yml:1]: Will be addressed as part of the following issue https://github.com/ocp-power-automation/ocp4-playbooks-extras/issues/54

**Review by yussufsh — APPROVED** [review]
/lgtm

**Pull request merged**